### ISSUE TITLE
Failure in setup function causes test to fail without displaying errors

### ISSUE THREAD

**Issue by cpmills1975** [bats-core/bats-core#72]
This may be my lack of understanding of bash, so forgive me if it is but I've come across the following situation which doesn't 'feel' right.

I've created a setup function which currently fails. All the setup function is configured to do is source a file which will contain constants but, at the moment, doesn't yet exist. Obviously this is a problem I will fix, but what concerned me was it resulted in the subsequent test failing but without displaying an error.

This can be replicated with the following test script:

```
function setup {
    source "/non-existant-file"
}

@test "some random test" {
    # This should always pass
    return 0
}
```

This will result in:

```
 ✗ some random test
```

which doesn't feel right. If I force the setup function to explicitly return 1:

```
function setup {
    source "/non-existant-file" || return 1
}
```

The result makes more sense:
```
 ✗ some random test
   (from function `setup' in test file test.bats, line 2)
     `source "/non-existant-file" || return 1' failed
   /tmp/bats.14184.src: line 2: /non-existant-file: No such file or directory
```
Is this my lack of understanding of bash or is something else at work here?

**Comment by cpmills1975** [comment]
It's probably worth mentioning that I installed bats from the EPEL 6 repo. Version number is 0.4.0 but the package file name (bats-0.4.0-1.20141016git3b33a5a.el6.noarch.rpm) appears to refer to git commit 3b33a5a made on 15th October 2014.

**Comment by mbland** [comment]
I've replicated your unexpected case with the original Bats 0.4.0, but the current bats-core master does exhibit the behavior you'd expect:

```bash
 ✗ some random test
   (from function `setup' in test file foo.bats, line 4)
     `source "/non-existent-file"' failed
   /var/folders/3z/7g_kcxy9693bh4ghtm9rtm7wgwxx_9/T/bats.9295.src: line 4: /non-existent-file: No such file or directory
```

The commit that fixed it was 0f6dde530ee41c36dcb32341ca41af49fb00ad7d. I'd thought the bug fixed by that commit was only a Bash 3.2.57(1)-release issue, but I just replicated your results even with the latest 4.4.19(1)-release. This implies that, in this case, the `ERR` trap isn't firing even for the latest version of Bash.

Will try to dig into this some more, but this gets us much closer to the actual reason.

**Comment by mbland** [comment]
Chapter and verse from https://tiswww.case.edu/php/chet/bash/CHANGES that might explain why 3.2.57(1)-release needed this fix as described in the commit message for 0f6dde530ee41c36dcb32341ca41af49fb00ad7d (I'll do a comparison of Bash versions later to verify):

```
This document details the changes between this version, bash-4.4-beta, and
the previous version, bash-4.4-alpha.

1.  Changes to Bash

k.  Fixed a bug that caused the ERR and RETURN traps to be unset if they were
    set in a shell function but unset previously.
```

I'm wondering if there's a similar condition with the `source` builtin lurking, as it all sorts of interesting `trap` semantics. From `man bash` for 4.4.19(1)-release:

```
source filename [arguments]
  ...
  If  the  -T option is enabled, source inherits any trap on DEBUG; if it is not, any DEBUG trap string is saved and restored around
  the call to source, and source unsets the DEBUG trap while it executes.  If -T is not set, and the sourced file changes the  DEBUG
  trap,  the  new  value  is  retained when source completes.
  ...
```

**Comment by eedwards-sk** [comment]
I'm on bash 3.2.57(1) (OSX), and seeing similar issue, although I don't even get proper output if I add an explicit return:

```
#!/usr/bin/env bats

setup() {
  source "nonexistent" || return 1
}

@test "hello world" {
  echo "hello world"
}
```

```
→ bats example.test
 ✗ hello world

1 test, 1 failure
```

vs

```
#!/usr/bin/env bats

setup() {
  false
}

@test "hello world" {
  echo "hello world"
}
```

```
→ bats example.test
 ✗ hello world
   (from function `setup' in test file example.test, line 4)
     `false' failed

1 test, 1 failure
```

**Comment by eedwards-sk** [comment]
I found a workaround by toggling `set -e` during the source command:

```
setup() {
  set +e
  source 'invalid'
  set -e
  return 1
}

@test "hello world" {
  echo "hello world"
}
```

```
→ ./example.test
 ✗ hello world
   (from function `setup' in test file example.test, line 5)
     `source 'invalid'' failed
   /var/folders/dm/2_68m_h13h5dxq2f05wsc47c0000gq/T/bats.75986.src: line 5: invalid: No such file or directory

1 test, 1 failure
```

**Comment by mbland** [comment]
This output would suggest that the ERR trap is firing as expected for the `source 'invalid'` line when error-on-exit is disabled, but that only the EXIT trap fires when it is.

So I'm not so sure the chapter-and-verse from [my earlier comment][earlier] applies. Still, something changed such that later versions of Bash fire the ERR trap before the EXIT trap. 

[earlier]: https://github.com/bats-core/bats-core/issues/72#issuecomment-379414539

One day I'll dig deeper into the Bash code to figure it out exactly—but really, the best workaround is to use the latest Bats from this repo, not the original 0.4.0 distribution. (And yes, I know we need to get #16 resolved...)

**Comment by eedwards-sk** [comment]
@mbland my example is from the latest version from this repo

**Comment by mbland** [comment]
Hmm, OK, I did just recreate that with 3.2.57(1)-release and the latest Bats. In fact, I've recreated the issue _without_ bats. The smoking gun is that:

* `source "nonexistent"` itself never triggers the ERR trap in any version, in any case (the ERR trap is only triggered by the nonzero return value of `do_it()`); and
* under 3.2.57(1)-release, with `set -e` enabled (i.e. error-on-exit), `source "nonexistent"` exits the script with a zero status immediately on failure. It returns nonzero without exiting in all other cases.

Here's my test file, `source.bash`:

```bash
#! /usr/bin/env bash

trap 'echo ERR:  status == "$?", funcname == "${FUNCNAME:-main}"' ERR
trap 'echo EXIT: status == "$?", funcname == "${FUNCNAME:-main}"' EXIT

do_it() {
  source "nonexistent" || return 1
}
do_it
```

In the following runs, `/bin/bash` is 3.2.57(1)-release, and `/usr/local/bin/bash` is 4.4.19(1)-release:

```
$ /bin/bash source.bash 
source.bash: line 7: nonexistent: No such file or directory
ERR: status == 1, funcname == main
EXIT: status == 1, funcname == main

$ /bin/bash -e source.bash 
source.bash: line 7: nonexistent: No such file or directory
EXIT: status == 0, funcname == do_it

$ /usr/local/bin/bash source.bash 
source.bash: line 7: nonexistent: No such file or directory
ERR: status == 1, funcname == main
EXIT: status == 1, funcname == main

$ /usr/local/bin/bash -e source.bash 
source.bash: line 7: nonexistent: No such file or directory
ERR: status == 1, funcname == main
EXIT: status == 1, funcname == main
```

Removing the `|| return 1` part of the expression produces:

```
$ /bin/bash source.bash 
source.bash: line 7: nonexistent: No such file or directory
ERR: status == 1, funcname == main
EXIT: status == 1, funcname == main

$ /bin/bash -e source.bash 
source.bash: line 7: nonexistent: No such file or directory
EXIT: status == 0, funcname == do_it

$ /usr/local/bin/bash source.bash 
source.bash: line 7: nonexistent: No such file or directory
ERR: status == 1, funcname == main
EXIT: status == 1, funcname == main

$ /usr/local/bin/bash -e source.bash 
source.bash: line 7: nonexistent: No such file or directory
EXIT: status == 1, funcname == do_it
```

I've found nothing in https://tiswww.case.edu/php/chet/bash/CHANGES that indicates when this behavior changed. Will have to dig into it later.

At any rate, I think I've convinced myself that it's worth suggesting `load()` over `source` in Bats tests, and toggling `set +e/-e` around the `source` command in its implementation, along with an explicit error check. I'll file an issue to that effect.

---

FWIW, here's the same matrix using Bats tests crafted from those in this thread:

Content of `source-nonexistent.test`:
```bash
#! /usr/bin/env bats

function setup {
  source "source nonexistent file"
}

@test "source nonexistent file" {
  :
}
```

Source of `source-nonexistent-and-return.test`:

```bash
#! /usr/bin/env bats

setup() {
  source "source nonexistent file and return 1" || return 1
}

@test "source nonexistent file and return 1" {
  :
}
```

Note that in the following cases, Bats uses several variables to detect when a test failed to complete, but uses an ERR trap to capture the error stack trace. This is why Bats reports legitimate failures, but with no stack information when the ERR trap doesn't fire.

Run with Bash 3.2.57(1)-release and original Bats 0.4.0 (`source` returns zero, `bats_error_trap` never executes):

```
$ PATH="/bin:$PATH" bats-orig/bin/bats source-nonexistent.test source-nonexistent-and-return.test 
 ✗ source nonexistent file
 ✗ source nonexistent file and return 1

2 tests, 2 failures
```

Run with Bash 3.2.57(1)-release and bats-core master at ffce883928419e2c3e31809bd42515d64672ad59 (`source` returns zero, `bats_error_trap` executes via `bats_teardown_trap` with `$?` always equal to zero):

```
$ PATH="/bin:$PATH" bats-core/bin/bats source-nonexistent.test source-nonexistent-and-return.test 
 ✗ source nonexistent file
 ✗ source nonexistent file and return 1

2 tests, 2 failures
```

Run with Bash 4.4.19(1)-release and original Bats 0.4.0 (`source` returns nonzero, `bats_error_trap` only executes on `|| return 1`):

```
$ PATH="/usr/local/bin:$PATH" bats-orig/bin/bats source-nonexistent.test source-nonexistent-and-return.test 
 ✗ source nonexistent file
 ✗ source nonexistent file and return 1
   (from function `setup' in test file source-nonexistent-and-return.test, line 4)
     `source "source nonexistent file and return 1" || return 1' failed
   /var/folders/3z/7g_kcxy9693bh4ghtm9rtm7wgwxx_9/T/bats.9802.src: line 4: source nonexistent file and return 1: No such file or directory

2 tests, 2 failures
```

Run with Bash 4.4.19(1)-release and bats-core master at ffce883928419e2c3e31809bd42515d64672ad59 (`source` returns nonzero, `bats_error_trap` always executes):

```
$ PATH="/usr/local/bin:$PATH" bats-core/bin/bats source-nonexistent.test source-nonexistent-and-return.test 
 ✗ source nonexistent file
   (from function `setup' in test file source-nonexistent.test, line 4)
     `source "source nonexistent file"' failed
   /var/folders/3z/7g_kcxy9693bh4ghtm9rtm7wgwxx_9/T/bats.9874.src: line 4: source nonexistent file: No such file or directory
 ✗ source nonexistent file and return 1
   (from function `setup' in test file source-nonexistent-and-return.test, line 4)
     `source "source nonexistent file and return 1" || return 1' failed
   /var/folders/3z/7g_kcxy9693bh4ghtm9rtm7wgwxx_9/T/bats.9886.src: line 4: source nonexistent file and return 1: No such file or directory

2 tests, 2 failures
```

**Comment by tterranigma** [comment]
In an effort to understand what is going on, I found [this bug report](https://lists.gnu.org/archive/html/help-bash/2015-11/msg00011.html) that might explain some part of the behavior above. In the [changelog](https://tiswww.case.edu/php/chet/bash/CHANGES), it is dealt with in the line:

```
o.  Fixed a bug that caused set -e to be honored in cases of builtins invoking
    other builtins when it should be ignored.
```

I will call the 2nd block of code in @mbland's comment above (the round with `|| return 1`) as A and the 3rd block (the round without `|| return 1`) as B. The four results in each will be numbered from 1 to 4. So B2 is:

```

$ /bin/bash -e source.bash 
source.bash: line 7: nonexistent: No such file or directory
EXIT: status == 0, funcname == do_it
```

In block A, `|| return 1` should effectively disable `set -e` (as mentioned in the link above). This means that A3 and A4 should and are equivalent. Since `source nonexistent` also returns with 1, B3 is also the same as A3 an A4. Now, the fact that A1 and A2 are not equivalent could be due to the bug I linked above which makes case A2 and B2 be equivalent when they shouldn't (the bug mentions nested builtins, could it be that the function declaration calling the `source` builtin is a case of nested builtins?).

Also, the fact that in these two cases (A2 and B2) the status is 0 although there was an error is probably due to one of the other numerous bugs related to traps, -e and statuses mentioned [in the changelog](https://tiswww.case.edu/php/chet/bash/CHANGES).

What I cannot understand is why in case B2 and B4 `set -e` stops the `ERR` trap from being executed. `source nonexistent` does indeed fire the ERR trap without -e, eg:

```
error_trap() {
  echo 'in error trap'
}
trap error_trap ERR
source hj
```

results in:

```
v: line 5: hj: No such file or directory
in error trap
```

Could this be a bug in Bash? Any possible explanation?

**Comment by mbland** [comment]
Just noticed this chapter and verse from https://tiswww.case.edu/php/chet/bash/CHANGES during my investigation of #81:

```
This document details the changes between this version, bash-4.1-alpha,
and the previous version, bash-4.0-release.

1.  Changes to Bash

x.  Fixed a bug that caused $? to not be set correctly when referencing an
    unset variable with set -u and set -e enabled.
```

And as I noted there:

> Now I'm wondering if that same fix might've affected the exit status of `source "nonexistent file"` as well. Will definitely have to take a look at the source and pinpoint the change when I get a chance. (I've got a system set up on my home machine to do this, but only have my work machine at the moment.)

**Comment by mbland** [comment]
While I pinpointed the exact cause and resolution for #81, I also ruled out that the same patch resolved this one. What I did find, using the same methodology of building different Bash versions, is that [Bash 4.2.23](https://ftp.gnu.org/gnu/bash/bash-4.2-patches/bash42-023) fixes it in a similar fashion.

Using the test script from before:

```bash
#! /usr/bin/env bash

trap 'echo ERR:  status == "$?", funcname == "${FUNCNAME:-main}"' ERR
trap 'echo EXIT: status == "$?", funcname == "${FUNCNAME:-main}"' EXIT

do_it() {
  source "nonexistent"
}
do_it
```

Produces:

```
$ 4.2.22/bin/bash -e source.test 
source.test: line 7: nonexistent: No such file or directory
EXIT: status == 0, funcname == do_it

$ 4.2.23/bin/bash -e source.test 
source.test: line 7: nonexistent: No such file or directory
EXIT: status == 1, funcname == do_it
```

More specifically, this part of the patch is what does the trick, I think:

```diff
*** ../bash-4.2-patched/error.c	2009-08-21 22:31:31.000000000 -0400
--- error.c	2012-02-25 15:54:40.000000000 -0500
***************
*** 201,205 ****
    va_end (args);
    if (exit_immediately_on_error)
!     exit_shell (1);
  }
  
--- 201,209 ----
    va_end (args);
    if (exit_immediately_on_error)
!     {
!       if (last_command_exit_value == 0)
! 	last_command_exit_value = 1;
!       exit_shell (last_command_exit_value);
!     }
  }
 
```
I still need to trace through why this fix works, which I think will answer why the ERR trap is never fired for `source` as well, as the code path for builtins is a little opaque. (@tterranigma I think the ERR trap will fire without `set -e` because a bad `source` command comprises a "pipeline" that fails, whereas with `set -e` it's a builtin failure that triggers a direct exit.) But at least we have the exact patch level wherein a `source` failure begins to report an error status.

**Comment by mbland** [comment]
OK, now I know exactly why `source "nonexistent file"` always exits without triggering the `ERR` trap, even in the latest Bash 4.4.19, and why the [Bash 4.2.23 patch](https://ftp.gnu.org/gnu/bash/bash-4.2-patches/bash42-023) fixes the exit status. The following logic traces through the Bash 4.4.19 source.

> Note that if you look through the code yourself, you'll see some conditions where other paths might execute if `POSIXLY_CORRECT` is in effect that do similar things, but bear in mind I'm describing what happens when `POSIXLY_CORRECT` is _not_ in effect.

The `source` builtin is defined in `builtins/source.def` and contains a call to `source_file()`.

> It searches for the filename before this call, but will continue to the `source_file()` call _unless_ `POSIXLY_CORRECT` is set, as that's what would cause `source_searches_cwd == 0` to trigger. See `posix_initialize()` in `general.c`.

`source_file()` is defined in `builtins/evalfile.c`, and just passes the `filename` through to `_evalfile()`, defined in the same file.

`_evalfile()` then calls `open (filename, O_RDONLY)`, and on error, calls `file_error (filename)`.

`file_error()` is a wrapper around `report_error ("%s: %s", filename, strerror (errno))`, which brings us to the 4.2.23 patch that updates `report_error` to set `last_command_exit_value = 1`:

```c
if (exit_immediately_on_error)
  {
    if (last_command_exit_value == 0)
      last_command_exit_value = 1;
    exit_shell (last_command_exit_value);
  }
```

`exit_shell()`, which is defined in `shell.c`, explicitly calls the `EXIT` trap, but not the `ERR` trap, before calling `exit()`.

So there, now we know why a failing `source` call by itself (i.e. when not followed by `|| return 1`) will bypass the `ERR` trap.

---

However, what's interesting is that the behavior of the `|| return 1` case was fixed somehow by 4.0, not by 4.0.23. Again, running the script:

```bash
#! /usr/bin/env bash

trap 'echo ERR:  status == "$?", funcname == "${FUNCNAME:-main}"' ERR
trap 'echo EXIT: status == "$?", funcname == "${FUNCNAME:-main}"' EXIT

do_it() {
  source "nonexistent" || return 1
}
do_it
```

produces (where `/bin/bash` is 3.2.57(1)-release):

```
$ /bin/bash -e source.test 
source.test: line 7: nonexistent: No such file or directory
EXIT: status == 0, funcname == do_it

$ 4.0/bin/bash -e source.test 
source.test: line 7: nonexistent: No such file or directory
ERR: status == 1, funcname == main
EXIT: status == 1, funcname == main
```

Since I'm in the groove, I built two of my own versions of 3.2.57, one from unaltered sources (to make sure I see the same behavior as with `/bin/bash`), and one with the following patch I divined from a diff against 4.0:

```diff
--- 3.2/bash-3.2.57/execute_cmd.c	2008-11-18 08:14:59.000000000 -0500
+++ 4.0/bash-4.0.0/execute_cmd.c	2009-02-13 16:41:41.000000000 -0500
@@ -3111,14 +3804,21 @@
   int old_e_flag, result, eval_unwind;
   int isbltinenv;
 
+#if 0
+  /* XXX -- added 12/11 */
+  terminate_immediately++;
+#endif
+
   old_e_flag = exit_immediately_on_error;
   /* The eval builtin calls parse_and_execute, which does not know about
      the setting of flags, and always calls the execution functions with
      flags that will exit the shell on an error if -e is set.  If the
      eval builtin is being called, and we're supposed to ignore the exit
      value of the command, we turn the -e flag off ourselves, then
-     restore it when the command completes. */
-  if (subshell == 0 && builtin == eval_builtin && (flags & CMD_IGNORE_RETURN))
+     restore it when the command completes.  This is also a problem (as
+     below) for the command and source/. builtins. */
+  if (subshell == 0 && (flags & CMD_IGNORE_RETURN) &&
+	(builtin == eval_builtin || builtin == command_builtin || builtin == source_builtin))
     {
       begin_unwind_frame ("eval_builtin");
       unwind_protect_int (exit_immediately_on_error);
```

Sure enough, it did the trick:

```
$ 3.2.57/bin/bash -e source.test 
source.test: line 7: nonexistent: No such file or directory
EXIT: status == 0, funcname == do_it

$ 3.2.57-with-builtin-patch/bin/bash -e source.test 
source.test: line 7: nonexistent: No such file or directory
ERR: status == 1, funcname == main
EXIT: status == 1, funcname == main
```

---

I have a few thoughts in mind pertaining to both this issue and #81 that I'll capture in a new issue.
